llm/rag/scripts/serve_rag.py

```diff
@@ -0,0 +1,568 @@
+First, explain your thinking process between <think> tags.
+Then provide your final answer after the thinking process.
+
+# Question:
+{query}
+
+Let's approach this step by step:"""
+
+
+async def query_llm(prompt: str, temperature: float = 0.7) -> tuple[str, str]:
+    """Query DeepSeek R1 through vLLM endpoint and return thinking process and answer."""
+    global generator_endpoint
+
+    try:
+        response = requests.post(f"{generator_endpoint}/v1/chat/completions",
+                                 json={
+                                     "model": "/tmp/generation_model",
+                                     "messages": [{
+                                         "role": "user",
+                                         "content": prompt
+                                     }],
+                                     "temperature": temperature,
+                                     "max_tokens": 2048,
+                                     "stop": None
+                                 },
+                                 timeout=120)
+        response.raise_for_status()
+
+        logger.info(f"Response: {response.json()}")
+
+        full_response = response.json(
+        )['choices'][0]['message']['content'].strip()
+
+        # Split response into thinking process and answer
+        parts = full_response.split("</think>")
+        if len(parts) > 1:
+            thinking = parts[0].replace("<think>", "").strip()
+            answer = parts[1].strip()
+        else:
+            thinking = ""
+            answer = full_response
+
+        return thinking, answer
+
+    except Exception as e:
+        logger.error(f"Error querying LLM: {str(e)}")
+        raise HTTPException(status_code=500,
+                            detail="Error querying language model")
+
+
+@app.post('/rag', response_model=RAGResponse)
+async def rag_query(request: QueryRequest):
+    """RAG endpoint combining vector search with DeepSeek R1."""
+    try:
+        # Encode query
+        query_embedding = encode_query(request.query)
+
+        # Get relevant documents
+        results = query_collection(query_embedding, request.n_results)
+
+        # Generate prompt
+        prompt = generate_prompt(request.query, results)
+
+        # Get LLM response
+        thinking, answer = await query_llm(prompt, request.temperature)
+
+        return RAGResponse(answer=answer,
+                           sources=results,
+                           thinking_process=thinking)
+
+    except Exception as e:
+        logger.error(f"Error processing RAG query: {str(e)}")
+        raise HTTPException(status_code=500, detail=str(e))
+
+
+@app.get('/health')
+async def health_check():
+    """Health check endpoint."""
+    return {
+        'status': 'healthy',
+        'collection_size': collection.count() if collection else 0
+    }
+
+
+@app.get('/', response_class=HTMLResponse)
+async def get_search_page():
+    """Serve a simple search interface."""
+    return """
```

Comment:
better to save the base html string in an additional file and read the file in the code here to make the code more modularized and clean?